### PR TITLE
Transfer by Dialpad

### PR DESCRIPTION
**Pull request by dbkr**
Adds a dial pad to the transfer dialog

Extends TabbedView to have a tabs-on-top view as well as tabs-on-left

The transfer dialog really needs to be split out of the invite dialog now, but that requires also splitting out the user search bit into its own component.

There are other general dial pad updates in the designs (adding letters, moving delete button to the text field) that I haven't done here because this PR is already sizeable.

Screenshot 2021-06-23 at 16 22 50
Screenshot 2021-06-23 at 16 22 44
Screenshot 2021-06-23 at 16 22 35
Screenshot 2021-06-23 at 16 22 28

### CODE REVIEW

File: src/components/views/dialogs/InviteDialog.tsx [hunk 1]
```diff
@@ -766,45 +779,32 @@ export default class InviteDialog extends React.PureComponent<IInviteDialogProps
...
+            const targetIds = targets.map(t => t.userId);
+            if (targetIds.length > 1) {
                 this.setState({
-                    busy: false,
-                    errorText: _t("Failed to transfer call"),
+                    errorText: _t("A call can only be transferred to a single user."),
                 });
             }
+
+            dis.dispatch({
```
Comment (gsouquet) [inline src/components/views/dialogs/InviteDialog.tsx:792]: I'm wondering whether this action should be called if `targetIds.length` is bigger than `1` ? You're setting an error but still transfering to the first of the two ids selected

File: src/components/views/dialogs/InviteDialog.tsx [hunk 2]
```diff
@@ -825,6 +825,10 @@ export default class InviteDialog extends React.PureComponent<IInviteDialogProps
         }
     };
 
+    private onCancel = () => {
+        this.props.onFinished([]);
```
Comment (gsouquet) [inline src/components/views/dialogs/InviteDialog.tsx:829]: If I remember correctly, the first parameter of this `onFinished` callback is a `boolean` to notify whether the modal closed successfully or in an error state

File: src/components/views/dialogs/InviteDialog.tsx [hunk 3]
```diff
@@ -1249,6 +1262,28 @@ export default class InviteDialog extends React.PureComponent<IInviteDialogProps
         }
     }
 
+    private onDialFormSubmit = ev => {
+        ev.preventDefault();
+        this.transferCall();
+    }
+
+    private onDialChange = ev => {
+        this.setState({dialPadValue: ev.target.value});
```
Comment (gsouquet) [inline src/components/views/dialogs/InviteDialog.tsx:1271]: I believe if you're setting `React.FormEvent` as the type of the parameter, TypeScript will complain that you're using `target` over `currentTarget`. As the latter is always guaranteed to be the correct element in all scenarios whether `target` depending on the event type and the context could be different

File: res/css/views/dialogs/_InviteDialog.scss [hunk 4]
```diff
@@ -316,3 +332,33 @@ limitations under the License.
 .mx_InviteDialog_helpText .mx_AccessibleButton_kind_link {
     padding: 0;
 }
+
+.mx_InviteDialog_dialPad .mx_InviteDialog_dialPadField {
+    border-top: 0px;
+    border-left: 0px;
+    border-right: 0px;
+    border-radius: 0px;
```
Comment (gsouquet) [inline res/css/views/dialogs/_InviteDialog.scss:340]: Probably out of scope for this review, but we might want to introduce a StyleLint rule that forces unit-less `0`.
It's what I usually see in CSS codebases

File: res/css/structures/_TabbedView.scss [hunk 5]
```diff
@@ -20,19 +21,99 @@ limitations under the License.
...
+        width: 22px;
+        height: 22px;
+        margin-left: 0px;
+        margin-right: 8px;
+    }
+
+    .mx_TabbedView_maskedIcon::before {
+        mask-size: 22px;
+        width: 22px;
+        height: 22px;
```
Comment (gsouquet) [inline res/css/structures/_TabbedView.scss:111]:
```suggestion
        width: inherit;
        height: inherit;
```

Also that does not solve the fact that the `mask-size` needs to match the parent icon width. Maybe using a CSS property here could help?

**Review by gsouquet — APPROVED** [review]
Otherwise looks good to me 👍

File: src/components/views/dialogs/InviteDialog.tsx (hunk 1 above)
Comment (dbkr) [reply src/components/views/dialogs/InviteDialog.tsx:792]: oops - you wonder correctly

File: src/components/views/dialogs/InviteDialog.tsx (hunk 2 above)
Comment (dbkr) [reply src/components/views/dialogs/InviteDialog.tsx:829]: This one appears to be special and unique and gets passed the list of strings to invite instead... :/

File: src/components/views/dialogs/InviteDialog.tsx (hunk 3 above)
Comment (dbkr) [reply src/components/views/dialogs/InviteDialog.tsx:1271]: TIL!

File: res/css/views/dialogs/_InviteDialog.scss (hunk 4 above)
Comment (dbkr) [reply res/css/views/dialogs/_InviteDialog.scss]: oh yeah - have changed these, but yep, we could probably add a rule

File: res/css/structures/_TabbedView.scss (hunk 5 above)
Comment (dbkr) [reply res/css/structures/_TabbedView.scss]: Yeah - I also think I'd managed to do these with less CSS before, although unsure if that made them more brittle or something. Maybe there's a case for a MaskedIcon component or something.

**Comment by dbkr** [comment]
@gaelledel design tweaks applied: I've update the first screenshot

**Comment by dbkr** [comment]
@gsouquet re-requesting review as quite as lot has changed. Plan is now to merge this to a branch and make more changes to keep the PRs a manageable size for code review but merge once we've achieved the final design.

**Review by gsouquet — APPROVED** [review]
Makes sense to me! :shipit:

**Pull request merged**